tests/test_hypothesis.py

```diff
@@ -0,0 +1,88 @@
+import os
+from datetime import datetime, timezone
+from typing import Optional
+
+import pytest
+from hypothesis import given, settings, strategies
+from typing_extensions import TypedDict
+
+from pydantic_core import SchemaValidator, ValidationError
+
+settings_dict = {}
+hyp_max_examples = os.getenv('HYPOTHESIS_MAX_EXAMPLES')
+if hyp_max_examples:
+    settings_dict['max_examples'] = int(hyp_max_examples)
+    print(f'hypothesis settings: {settings_dict}')
+
+
+@pytest.fixture(scope='module')
+def datetime_schema():
+    return SchemaValidator({'type': 'datetime'})
+
+
+@settings(**settings_dict)
+@given(strategies.datetimes())
+def test_datetime_datetime(datetime_schema, data):
+    assert datetime_schema.validate_python(data) == data
+
+
+@settings(**settings_dict)
+@given(strategies.integers(min_value=-11_676_096_000, max_value=253_402_300_799_000))
+def test_datetime_int(datetime_schema, data):
+    if abs(data) > 20_000_000_000:
+        microsecond = (data % 1000) * 1000
+        expected = datetime.fromtimestamp(data // 1000, tz=timezone.utc).replace(tzinfo=None, microsecond=microsecond)
+    else:
+        expected = datetime.fromtimestamp(data, tz=timezone.utc).replace(tzinfo=None)
+
+    assert datetime_schema.validate_python(data) == expected, data
+
+
+@settings(**settings_dict)
+@given(strategies.binary())
+def test_datetime_binary(datetime_schema, data):
+    try:
+        datetime_schema.validate_python(data)
+    except ValidationError:
+        # that's fine
+        pass
+
+
+@pytest.fixture(scope='module')
+def recursive_schema():
+    return SchemaValidator(
+        {
+            'type': 'typed-dict',
+            'ref': 'Branch',
+            'fields': {
+                'name': {'schema': {'type': 'str'}},
+                'sub_branch': {
+                    'schema': {'type': 'nullable', 'schema': {'type': 'recursive-ref', 'schema_ref': 'Branch'}},
+                    'default': None,
+                },
+            },
+        }
+    )
+
+
+def test_recursive_simple(recursive_schema):
+    assert recursive_schema.validate_python({'name': 'root'}) == {'name': 'root', 'sub_branch': None}
+
+
+class BranchModel(TypedDict):
+    name: str
+    sub_branch: Optional['BranchModel']
+
+
+@settings(**settings_dict)
+@given(strategies.from_type(BranchModel))
+def test_recursive_hyp(recursive_schema, data):
```

Comment:
The reason that this test passes is that Hypothesis *doesn't* generate cyclic references by default; unless you write a strategy which can explicitly choose between "generate a new BranchModel" and "reuse an existing BranchModel" you'll always get the former.  That might look something like:

```python
@st.composite
def branch_models_with_cycles(draw, existing: list[BranchModel] | None = None) -> BranchModel:
    if existing is None:
        existing = []
    name = draw(st.text())
    sub_branch = draw(st.builds(BranchModel) | st.sampled_from(existing))  # might need to handle empty-list case?
    return BranchModel(name=name, sub_branch=sub_branch)
```

Reply:
Looks good, I'll add this.

Reply:
I've tried tweaking this and got a recursion error in hypothesis (I think)

Adjusted function:

```py
@strategies.composite
def branch_models_with_cycles(draw, existing: list[BranchModel] = strategies.from_type(list[BranchModel])):
    # debug(existing)
    name = draw(strategies.text())
    sub_branch = draw(strategies.from_type(BranchModel) | strategies.sampled_from(existing))
    branch = BranchModel(name=name, sub_branch=sub_branch)
    existing.append(branch)
    return branch
```

The traceback is massive, [gist of the output here](https://gist.github.com/samuelcolvin/639af81fe0eaa211ee05bf7f4ec46354).

Reply:
This one works:
```python

@st.composite
def branch_models_with_cycles(draw, existing: list[BranchModel] | None = None):
    if existing is None:
        existing = []
    name = draw(st.text())
    sub_branch = draw(
        st.none()
        | (st.sampled_from(existing) if existing else st.nothing())
        | branch_models_with_cycles(existing)
    )
    branch = BranchModel(name=name, sub_branch=sub_branch)
    existing.append(branch)
    return branch
```

I think your underlying problem was that `existing: list[BranchModel] = strategies.from_type(list[BranchModel])` is ill-typed; the default value is a *strategy which generates* `list[BranchModel]`, rather than an instance of that type.

Reply:
I've tried that unfortunately it's not causing a failure, since the base of this PR pre-dates #134, I think that means this code still isn't generate cyclic references.

(Ignore the failures on CI, that's just rust nightly and will fix itself soon).

Reply:
🤦 yeah, by the clear light of day we have to append to the `existing` list *before* we draw from it, not afterwards, and then mutate the `sub_branch` in place afterwards: 

```python
@st.composite
def branch_models_with_cycles(draw, existing=None):
    if existing is None:
        existing = []
    model = BranchModel(name=draw(st.text()), sub_branch=None)
    existing.append(model)
    model["sub_branch"] = draw(
        st.none()
        | st.builds(BranchModel, name=st.text(), sub_branch=branch_models_with_cycles(existing))
        | st.sampled_from(existing)
    )
    return model
```

I've checked that this generates cycles as well as just generating values at all.  Sorry for the back-and-forth!

Reply:
Not at all! I should be the one apologising.

Thanks so much, that's working now.